keras_nlp/samplers/sampler.py

```diff
@@ -0,0 +1,279 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Base sampler class."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+@keras.utils.register_keras_serializable(package="keras_nlp")
+class Sampler:
+    """Base sampler class.
+
+    Args:
+        {{base_optimizer_keyword_args}}
+
+    Call Args:
+        {{call_keyword_docstring}}
+
+    The inputs and outputs of Sampler class are both token ids.
+
+    Examples:
+
+    Basic usage:
+    ```python
+    BATCH_SIZE = 8
+    VOCAB_SIZE = 10
+    FEATURE_SIZE = 16
+    START_ID = 1
+    END_ID = 2
+
+    # Create a dummy model to predict the next token.
+    model = keras.Sequential(
+        [
+            keras.Input(shape=[None]),
+            keras.layers.Embedding(
+                input_dim=VOCAB_SIZE,
+                output_dim=FEATURE_SIZE,
+            ),
+            keras.layers.Dense(VOCAB_SIZE, activation="softmax"),
+        ]
+    )
+
+    # Define a function that outputs the next token's probability for each token
+    # in the input sequence.
+    def token_probability_fn(inputs, mask):
+        return model(inputs)
+
+    prompt = tf.fill((BATCH_SIZE, 1), START_ID)
+
+    sampler = keras_nlp.samplers.Greedy()
```

Comment:
This is documenting the `Greedy` class on the base class, which seems odd. Maybe leave these examples off, and show the detokenization example @jbischof asked for on the actual `Greedy` class?

Reply:
Good point. Now that we're introducing a lot of abstract base classes (ones no one should use directly), we may want to start noting at the top of docstrings e.g., "An abstract base class for samplers." I could imagine the same for `Backbone` and `Preprocessor`